### PR TITLE
Finish quoting

### PR DESCRIPTION
**Pull request by penev92**
Finalized the quoting by adding a bunch more quoting methods and enhancements, as well as fixing and refactoring the code.
Most of the new code was originally written for [Orabot](https://github.com/OpenRA/Orabot) and I am back-porting it to SteakBot. So the commits here aren't as well structured. With that you might want to review it there. Once it's merged here I'll merge it there as well.

### CODE REVIEW

File: SteakBot.Core/Services/QuotingService.cs [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+            var firstMessage = await channel.GetMessageAsync(firstMessageId);
+            if (firstMessage == null)
+            {
+                return new List<IMessage>();
+            }
+
+            var downloadedMessages = channel.GetMessagesAsync(firstMessageId, Direction.After, 25);
+            var firstPage = (await downloadedMessages.First()).ToArray();
+
+            var messages = new List<IMessage>
```
Comment (naceto) [inline SteakBot.Core/Services/QuotingService.cs]: You can specify the size of the List collection `firstPage.Length  + 1` (+1 is for the `firstMessage`) that is added in the initializer if its not included in the `firstPage.Length`.

**Review by naceto — COMMENTED** [review]
Please streamline the method parameters in `HelpModule.cs` to use IEnumerable and remove the `.ToArray()` calls.

File: SteakBot.Core/Services/QuotingService.cs (hunk 1 above)
Comment (penev92) [reply SteakBot.Core/Services/QuotingService.cs]: Done.

**Comment by penev92** [comment]
Resolved 2 comments that were discussed - the List size one from above and the one about having 3 `out` parameters on a method.

**Review by naceto — APPROVED** [review]

**Review by naps250 — APPROVED** [review]
very changes
much good
wow

**Pull request merged**